### ISSUE TITLE
Publishing results of eager tasks

### ISSUE THREAD

**Issue by dtheodor** [celery/celery#2972]
Why are eager task results not published to the task backend? https://github.com/celery/celery/blob/816ff0310296aad7d44cd6e07004f770c007e226/celery/app/trace.py#L275

I don't think this is a good design choice. It disrupts the code flow, requiring to add more logic to handle eager results. For example, my code will do something like this:

```python
async_result = my_task.delay()
handle_task_result(async_result.id)

def handle_task_result(id):
  # do stuff with the result, either synchronously or asynchronously
  # the result can only be accessed through `AsyncResult(id)`
```

This won't work at all when eager is turned on; trying `AsyncResult(id)` within `handle_task_result` will return non existing 'PENDING' garbage. Instead, I have to add a workaround in the calling code

```python
async_result = my_task.delay()
if async_result.is_ready():
  # this is entered only when celery eager is on
  # forces to deal with the result synchronously
else:
  handle_task_result(async_result.id)
```

This additional body of code is 1. totally useless under regular circumstances, it only has value when debugging with eager on and 2. ugly, since the two ways to get results use a totally different API. If eager task results were published instead, the application code would not need to be modified at all.

Thoughts?

**Comment by ask** [comment]
Always eager is used for development when you don't want to bother installing a broker, so what do you do with installations where the result backend and broker transport is the same?  If using a database how do you handle cleanup?

Always eager is also frequently used in unit tests (even though that is now discouraged), and accessing
a database or any other external service invalidates the core principles of unit testing.

If the code calling the task and the code handling the result is executed in the same process, why
don't you just pass the result instance?

**Comment by dtheodor** [comment]
>Always eager is used for development when you don't want to bother installing a broker, so what do you do with installations where the result backend and broker transport is the same? If using a database how do you handle cleanup?

I use always eager for development; in particular I use always eager to be able to debug a task, i.e. step into it with the debugger I am already using for the calling process.

As for cleanup, 1. the dev db is being cleaned up by the clean up task; my dev broker is running, and 2. I don't think properly cleaning up a dev database is important at all at the state of running with always eager.

> If the code calling the task and the code handling the result is executed in the same process

That's the thing, this *never* happens *except* when I am debugging. I don't want to add more code to handle the synchronous result just for when I am debugging; this invalidates the flow that handles the asynchronous result which I am also debugging.

**Comment by dtheodor** [comment]
This is a duplicate of https://github.com/celery/celery/issues/2275

**Comment by ask** [comment]
Closed with #2275

**Comment by WilliamDEdwards** [comment]
https://github.com/celery/celery/issues/2275 doesn't actually solve the intent of this issue: storing the results. https://github.com/celery/celery/pull/6614 does.
